File: src/main/java/com/jackmeng/stl/stl_ZipFileReader.java
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package com.jackmeng.stl;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class stl_ZipFileReader
{
    private final Map<String, byte[]> cache;

    public stl_ZipFileReader(InputStream inputStream)
            throws IOException
    {
        cache = new HashMap<>();
        try (ZipInputStream zipInputStream = new ZipInputStream(inputStream))
        {
            ZipEntry entry;
            while ((entry = zipInputStream.getNextEntry()) != null)
            {
                byte[] content = zipInputStream.readAllBytes();
                cache.put(entry.getName(), content);
            }
        }
    }

    public InputStream getStream(String entryName)
    {
        byte[] content = cache.get(entryName);
        return content != null ? new ByteArrayInputStream(content) : null;
    }
}

File: src/main/java/com/jackmeng/stl/stl_ZipFileReader.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

import java.util.HashMap
import java.util.zip.ZipInputStream
import java.util.zip.ZipEntry
import java.io.ByteArrayInputStream
import java.io.InputStream

class stl_ZipFileReader(inputStream:InputStream?)
{
	private val cache:MutableMap<String , ByteArray>
	
	init
	{
		cache=HashMap()
		if (inputStream!=null)
		{
			ZipInputStream(inputStream).use { zipInputStream->
				var entry:ZipEntry
				while (zipInputStream.nextEntry.also { entry=it }!=null)
				{
					val content=zipInputStream.readAllBytes()
					cache[entry.name]=content
				}
			}
		}
	}
	
	fun getStream(entryName:String):InputStream?
	{
		val content=cache[entryName]
		return content?.let { ByteArrayInputStream(it) }
	}
}